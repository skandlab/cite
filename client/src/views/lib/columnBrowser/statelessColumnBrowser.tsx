import React from "react";
import { SemanticShorthandContent } from "semantic-ui-react";
import { BaseColumnBrowser, ColumnBrowserType } from "./baseColumnBrowser";

interface ColumnBrowserProps {
	title: string;
	options: ColumnBrowserType[];
	popupContent: SemanticShorthandContent;
	handleToggleCheckbox: (changedOptions: ColumnBrowserType[]) => void;
}

export const StatelessColumnBrowser = (props: ColumnBrowserProps) => {
	const handleToggleCheckbox = (option: ColumnBrowserType, index: number) => {
		let filteredOptionList = [];
		if (option.isChecked) {
			filteredOptionList = [
				...props.options.slice(0, index),
				...props.options.slice(index + 1, totalChecked),
			];
		} else {
			filteredOptionList = props.options.slice(0, totalChecked);
			filteredOptionList.push({ ...option, isChecked: true });
		}
		props.handleToggleCheckbox(filteredOptionList);
	};

	let totalChecked = props.options.filter((option) => option.isChecked)
		.length;

	return (
		<BaseColumnBrowser
			{...props}
			totalChecked={totalChecked}
			handleToggleCheckbox={handleToggleCheckbox}
			handleReset={() => props.handleToggleCheckbox([])}
		/>
	);
};
