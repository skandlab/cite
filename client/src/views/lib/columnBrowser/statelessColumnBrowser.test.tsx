import React from "react";
import "@testing-library/dom";
import "@testing-library/jest-dom/extend-expect";
import { render, cleanup, fireEvent } from "@testing-library/react";
import { StatelessColumnBrowser } from "./statelessColumnBrowser";
import {
	TestState,
	getFilterInput,
	getCheckboxToggle,
	getResetButton,
} from "./__utils__";

const optionValueList = ["cc", "cs", "nn"];

const mockedHandleToggleCheckbox = jest.fn((data) => null);

const options = [
	{
		index: 0,
		value: optionValueList[0],
		mute: false,
		isChecked: false,
	},
	{
		index: 1,
		value: optionValueList[1],
		mute: false,
		isChecked: false,
	},
	{
		index: 2,
		value: optionValueList[2],
		mute: false,
		isChecked: false,
	},
];

const InitialRender = () =>
	render(
		<StatelessColumnBrowser
			title="Interaction type"
			options={options}
			popupContent={
				<p>if no associated heatmap present then gene will be dimmed</p>
			}
			handleToggleCheckbox={mockedHandleToggleCheckbox}
		/>
	);

const OneCheckedRender = () =>
	render(
		<StatelessColumnBrowser
			title="Interaction type"
			options={[
				{
					index: 0,
					mute: true,
					isChecked: false,
					value: "cc",
					description: "cancer-cancer",
				},
				{
					index: 1,
					mute: true,
					isChecked: false,
					value: "cs",
					description: "cancer-stroma",
				},
				{
					index: 2,
					mute: false,
					isChecked: true,
					value: "nn",
					description: "normal-normal",
				},
			]}
			popupContent={null}
			handleToggleCheckbox={mockedHandleToggleCheckbox}
		/>
	);

/**
 * checked items will always be on top
 */
const TwoCheckedRender = () =>
	render(
		<StatelessColumnBrowser
			title="Interaction type"
			options={[
				{
					index: 0,
					mute: false,
					isChecked: true,
					value: "cc",
				},
				{
					index: 2,
					mute: false,
					isChecked: true,
					value: "nn",
				},
				{
					index: 1,
					mute: true,
					isChecked: false,
					value: "cs",
				},
			]}
			popupContent={null}
			handleToggleCheckbox={mockedHandleToggleCheckbox}
		/>
	);

afterEach(() => {
	cleanup();
	mockedHandleToggleCheckbox.mockClear();
});

test("initial render", () => {
	const { asFragment } = InitialRender();
	expect(asFragment()).toMatchSnapshot();

	TestState(
		"",
		[
			{ item: "cc", checked: false },
			{ item: "cs", checked: false },
			{ item: "nn", checked: false },
		],
		"0 / 3",
		true
	);
});

test("one item is checked", () => {
	const { asFragment } = OneCheckedRender();
	expect(asFragment()).toMatchSnapshot();

	TestState(
		"",
		[
			{ item: "cc", checked: false },
			{ item: "cs", checked: false },
			{ item: "nn", checked: true },
		],
		"1 / 3",
		false
	);
});

test("toggle", () => {
	InitialRender();

	/**
	 * initial render
	 */
	TestState(
		"",
		[
			{ item: "cc", checked: false },
			{ item: "cs", checked: false },
			{ item: "nn", checked: false },
		],
		"0 / 3",
		true
	);

	/**
	 * toggle nn
	 */
	fireEvent.click(getCheckboxToggle("nn"));
	expect(mockedHandleToggleCheckbox.mock.calls[0][0]).toEqual([
		{ ...options[2], isChecked: true },
	]);
	mockedHandleToggleCheckbox.mockClear();
});

test("two item is checked -> untoggle", () => {
	TwoCheckedRender();

	/**
	 * initial render
	 */
	TestState(
		"",
		[
			{ item: "cc", checked: true },
			{ item: "cs", checked: false },
			{ item: "nn", checked: true },
		],
		"2 / 3",
		false
	);

	/**
	 * toggle nn
	 */
	fireEvent.click(getCheckboxToggle("nn"));
	expect(mockedHandleToggleCheckbox.mock.calls[0][0]).toEqual([
		{ ...options[0], isChecked: true },
	]);
	mockedHandleToggleCheckbox.mockClear();
});

test("one item is checked -> reset", () => {
	OneCheckedRender();

	/**
	 * initial render
	 */
	TestState(
		"",
		[
			{ item: "cc", checked: false },
			{ item: "cs", checked: false },
			{ item: "nn", checked: true },
		],
		"1 / 3",
		false
	);

	/**
	 * reset
	 */
	fireEvent.click(getResetButton());
	expect(mockedHandleToggleCheckbox.mock.calls[0][0]).toEqual([]);
	mockedHandleToggleCheckbox.mockClear();
});

test("one item is checked -> filter -> reset", () => {
	OneCheckedRender();

	/**
	 * initial render
	 */
	TestState(
		"",
		[
			{ item: "cc", checked: false },
			{ item: "cs", checked: false },
			{ item: "nn", checked: true },
		],
		"1 / 3",
		false
	);

	/**
	 * filter "n"
	 */
	fireEvent.change(getFilterInput(), { target: { value: "n" } });
	TestState("n", [{ item: "nn", checked: true }], "1 / 3", false);

	/**
	 * reset
	 */
	fireEvent.click(getResetButton());
	expect(mockedHandleToggleCheckbox.mock.calls[0][0]).toEqual([]);
	mockedHandleToggleCheckbox.mockClear();
});

test("filter -> toggle", () => {
	/**
	 * initial
	 */
	InitialRender();
	TestState(
		"",
		[
			{ item: "cc", checked: false },
			{ item: "cs", checked: false },
			{ item: "nn", checked: false },
		],
		"0 / 3",
		true
	);

	/**
	 * input change
	 */
	fireEvent.change(getFilterInput(), { target: { value: "c" } });
	TestState(
		"c",
		[
			{ item: "cc", checked: false },
			{ item: "cs", checked: false },
		],
		"0 / 3",
		true
	);

	/**
	 * toggle
	 */
	fireEvent.click(getCheckboxToggle("cc"));
	expect(mockedHandleToggleCheckbox.mock.calls[0][0]).toEqual([
		{ ...options[0], isChecked: true },
	]);
	mockedHandleToggleCheckbox.mockClear();
});
