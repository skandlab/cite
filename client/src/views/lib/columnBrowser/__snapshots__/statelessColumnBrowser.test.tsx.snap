// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial render 1`] = `
<DocumentFragment>
  <div
    class="ui card"
  >
    <div
      class="content"
    >
      <div
        class="meta css-106mbhu"
      >
        <span>
          <i
            aria-hidden="true"
            class="filter icon"
          />
           Interaction type
        </span>
        <i
          aria-hidden="true"
          class="info circle icon"
        />
      </div>
      <div
        class="description"
      >
        <form
          class="ui form css-1rrhbx0"
        >
          <div
            class="field"
          >
            <div
              class="ui left icon input"
              data-testid="filterInput"
            >
              <input
                type="text"
                value=""
              />
              <i
                aria-hidden="true"
                class="search icon"
              />
            </div>
          </div>
        </form>
        <div
          class="ui items css-p5nlxd"
          data-testid="checkbox-list"
        >
          <div
            class="item css-1l92o6s"
          >
            <div
              class="content"
            >
              <div
                class="description"
              >
                <div
                  class="ui checkbox"
                  data-testid="cc"
                >
                  <input
                    class="hidden"
                    readonly=""
                    tabindex="0"
                    type="checkbox"
                    value=""
                  />
                  <label>
                    cc
                  </label>
                </div>
              </div>
            </div>
          </div>
          <div
            class="item css-1l92o6s"
          >
            <div
              class="content"
            >
              <div
                class="description"
              >
                <div
                  class="ui checkbox"
                  data-testid="cs"
                >
                  <input
                    class="hidden"
                    readonly=""
                    tabindex="0"
                    type="checkbox"
                    value=""
                  />
                  <label>
                    cs
                  </label>
                </div>
              </div>
            </div>
          </div>
          <div
            class="item css-1l92o6s"
          >
            <div
              class="content"
            >
              <div
                class="description"
              >
                <div
                  class="ui checkbox"
                  data-testid="nn"
                >
                  <input
                    class="hidden"
                    readonly=""
                    tabindex="0"
                    type="checkbox"
                    value=""
                  />
                  <label>
                    nn
                  </label>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="extra content"
    >
      <button
        class="ui mini icon disabled right labeled button"
        data-testid="resetButton"
        disabled=""
        tabindex="-1"
      >
        <i
          aria-hidden="true"
          class="undo icon"
        />
        Reset Filter
      </button>
      <div
        class="ui right floated statistic xs"
        data-testid="statistic"
      >
        <div
          class="value"
        >
          0 / 3
        </div>
        <div
          class="label"
        >
          showing
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`one item is checked 1`] = `
<DocumentFragment>
  <div
    class="ui card"
  >
    <div
      class="content"
    >
      <div
        class="meta css-106mbhu"
      >
        <span>
          <i
            aria-hidden="true"
            class="filter icon"
          />
           Interaction type
        </span>
      </div>
      <div
        class="description"
      >
        <form
          class="ui form css-1rrhbx0"
        >
          <div
            class="field"
          >
            <div
              class="ui left icon input"
              data-testid="filterInput"
            >
              <input
                type="text"
                value=""
              />
              <i
                aria-hidden="true"
                class="search icon"
              />
            </div>
          </div>
        </form>
        <div
          class="ui items css-p5nlxd"
          data-testid="checkbox-list"
        >
          <div
            class="item css-123mvh4"
            mute=""
          >
            <div
              class="content"
            >
              <div
                class="description"
              >
                <div
                  class="ui checkbox"
                  data-testid="cc"
                >
                  <input
                    class="hidden"
                    readonly=""
                    tabindex="0"
                    type="checkbox"
                    value=""
                  />
                  <label>
                    cc
                  </label>
                </div>
              </div>
              <div
                class="extra css-fv2hk0"
              >
                <small>
                  cancer-cancer
                </small>
              </div>
            </div>
          </div>
          <div
            class="item css-123mvh4"
            mute=""
          >
            <div
              class="content"
            >
              <div
                class="description"
              >
                <div
                  class="ui checkbox"
                  data-testid="cs"
                >
                  <input
                    class="hidden"
                    readonly=""
                    tabindex="0"
                    type="checkbox"
                    value=""
                  />
                  <label>
                    cs
                  </label>
                </div>
              </div>
              <div
                class="extra css-fv2hk0"
              >
                <small>
                  cancer-stroma
                </small>
              </div>
            </div>
          </div>
          <div
            class="item css-1l92o6s"
          >
            <div
              class="content"
            >
              <div
                class="description"
              >
                <div
                  class="ui checked checkbox"
                  data-testid="nn"
                >
                  <input
                    checked=""
                    class="hidden"
                    readonly=""
                    tabindex="0"
                    type="checkbox"
                    value=""
                  />
                  <label>
                    nn
                  </label>
                </div>
              </div>
              <div
                class="extra css-fv2hk0"
              >
                <small>
                  normal-normal
                </small>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="extra content"
    >
      <button
        class="ui mini icon right labeled button"
        data-testid="resetButton"
      >
        <i
          aria-hidden="true"
          class="undo icon"
        />
        Reset Filter
      </button>
      <div
        class="ui right floated statistic xs"
        data-testid="statistic"
      >
        <div
          class="value"
        >
          1 / 3
        </div>
        <div
          class="label"
        >
          showing
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
